### DISCUSSION TITLE
ParserError: Source file requires different compiler version

### DISCUSSION THREAD

**Discussion by gorkemkeles** [smartcontractkit/full-blockchain-solidity-course-py#522] [Q&A] [answered]
I started to get compile error after adding openzepplin ownable contract. I'm really confused and can't find solution.

Could you please someone help me about it?

![Screen Shot 2021-12-08 at 17 19 01](https://user-images.githubusercontent.com/18116965/145224791-093e6b49-5daa-4c55-a5ab-d4830e6412fd.png)

![Screen Shot 2021-12-08 at 17 31 25](https://user-images.githubusercontent.com/18116965/145226022-b2c821ea-ccf2-4ab6-9d88-4e0cd19a1b3b.png)

![Screen Shot 2021-12-08 at 17 19 33](https://user-images.githubusercontent.com/18116965/145224786-f59512ee-a2e5-4bb7-b75f-edf7e7ddc934.png)

**Comment by cromewar** [comment, marked as answer]
Hello @gorkemkeles it happens when the `pragma solidity` is not the same as the one your compiler is using, however I wonder which vscode extension are you using?  The Solidity extension by Juan Blanco allows to compile different versions as far the one you are using is superior than the required.

also change this:

`solidity >=0.6.0 <0.9.0`

to this

`solidity ^0.6.0`

Finally if you have linting warning you could refer to [this](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/blob/main/chronological-issues-from-video.md#linting-issues) apart.

![image](https://user-images.githubusercontent.com/20199180/145226217-b7d0d460-a223-4d78-9e10-d037f1c85125.png)

**Reply by gorkemkeles** [reply]
I changed solidity ^0.6.0 still getting an error like this:

![Screen Shot 2021-12-08 at 18 02 57](https://user-images.githubusercontent.com/18116965/145231505-62d47438-2b32-4d2a-ae86-b41ea20dde03.png)

I'm using the same vscode extension: 

![Screen Shot 2021-12-08 at 18 01 41](https://user-images.githubusercontent.com/18116965/145231588-fa80edbf-cda5-455e-95cc-0f2a1edae434.png)

**Comment by cromewar** [comment]
Look your current compiler version is this one:
![image](https://user-images.githubusercontent.com/20199180/145232514-8154ccb9-e129-4c5b-8715-93e45fa99ba8.png)
And as I said it would work with different versions as far as the one you are using is equal or superior, so on any contract you have opened on vscode, push right click and change it to 0.8.0
![image](https://user-images.githubusercontent.com/20199180/145232063-d0534139-a5b6-4c17-b7ef-dae1f4c881f8.png)
![image](https://user-images.githubusercontent.com/20199180/145232434-1140d3f4-26ca-44e1-abf0-095e1aa9bbdc.png)

**Reply by cromewar** [reply]
Also make sure this options is marked as remote:
![image](https://user-images.githubusercontent.com/20199180/145232754-5d0eb962-a43a-49e9-9b54-27daf1ec0b71.png)

**Reply by gorkemkeles** [reply]
current compiler version changed after changed to solidity ^0.6.0 it was 0.8.10 when it was solidity >=0.6.0 <0.9.0; as you can see in the first screenshot.

Also, I did, what did you say in your last answer unfortunately I'm getting the same error

![Screen Shot 2021-12-08 at 18 02 57](https://user-images.githubusercontent.com/18116965/145233943-852fa50d-fbb8-4784-8810-e33d64394d82.png)

**Reply by cromewar** [reply]
It says your compiler is 0.6.12, so its not taking the 0.8.0 change, try this.

1. delete the build folder on your project.
2. make sure to select remote on the comment I posted above.
3. close your vscode and all the terminals you have opened.
4. Open vscode and try again.

**Reply by gorkemkeles** [reply]
I did this but got the same error again my friend, before adding openzepplin to  depencies  on brownie-config.yaml everything was perfect, I really don't understand what did I wrong

**Reply by cromewar** [reply]
may be you could delete this folder: `.brownie/packages/OpenZeppelin` (this folder is on your home Path), also choose this option: 
![image](https://user-images.githubusercontent.com/20199180/145241720-e87d4e48-5521-4627-9faa-be3cdf83b039.png)
and also change it to 0.8.0 or superior.

**Reply by gorkemkeles** [reply]
Couldn't find the folder as OpenZeppelin, was it in the build folder that I deleted earlier?

**Reply by cromewar** [reply]
Nope that folder appears at compiling dependencies on brownie-config (if you are using MAC or Linux this is a hidden folder on your home path) anyway did you change the workspace compiler version?

**Reply by gorkemkeles** [reply]
yes, I did change but the same :) Sorry man, I feel, I'm wasting your time.

**Reply by gorkemkeles** [reply]
Is it possible to have this problem because of changes of v0.8 solidity about creating smart contract or function in it because in the video I think the version is v0.6 
for example, I need to use **payable** different than how Patrick doing in the course because I use a different version of solidity.

**Reply by cromewar** [reply]
never mind I'm here to help, I reviewed again the error code at it seems now the one complaining is the chainlink one as you can see bellow:

![image](https://user-images.githubusercontent.com/20199180/145249202-3d5ed988-5dbf-4382-b055-4e0cacdf88e4.png)


![image](https://user-images.githubusercontent.com/20199180/145248846-73c47626-4687-46e4-80dc-adb8253ef424.png)

So may be try to explicitly use 0.8.0

As a you have already changed the version both form global and workspace to 0.8.0 and selected remote.

Try to change this:

`pragma solidity ^0.6.0`

to this 

`pragma solidity 0.8.0`

this should work.

**Reply by gorkemkeles** [reply]
I already tried that but also not working ^0.0.8, the funny part is that I didn't get an error while using **>=0.6.0 <0.9.0;** before but now I get errors when using it either

This error when I change to ^0.0.8:

![Screen Shot 2021-12-08 at 19 51 55](https://user-images.githubusercontent.com/18116965/145249570-e5f6d315-32e2-4ef3-a923-91e0e5bf4ecb.png)
:

**Reply by cromewar** [reply]
I think I finally managed to locate the real problem:
well problem is that Openzeppelin ownable is expecting a version less than 0.8.0 and your compiler is using 0.8.10, which is strange as you can see on Openzeppelin contracts [github](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol) they are compatible indeed with this version. So I checked up again your `brownie-config` and this is the problem:

![image](https://user-images.githubusercontent.com/20199180/145254080-5ed42100-7974-466e-9d24-adcc297a261c.png)
 is not @3.4.0 is actually @4.3.0  (you could also use 4.3.2 or 4.4.0 which are the latest)

Give it a try ;)

**Reply by gorkemkeles** [reply]
YES! This is worked, man thank you so so much for your help and your time. It was 3.4.0 in the course video and I just used this version.

**Reply by cromewar** [reply]
This happens as the versions are constantly updating, I'm glad your problem is solved now.
I'll close this thread now and please feel free to make any other questions during the course.

Best regards, Cromewar.

**Reply by gorkemkeles** [reply]
Okay, thank you a lot again

**Reply by WeTheArtist** [reply]
HI @cromewar , I tried all that you mentioned in this discussion, however it is not solving the issue for me, I also noticed that the current compiler version remains (current compiler is 0.6.12+commit.27d51765.Windows.msvc). But when i right click and select Solidty:Compiler information, it shows 0.8.0.
![image](https://user-images.githubusercontent.com/79281462/147146252-07a6aeab-64d6-4e2f-839f-24675d9d974e.png)

```
Retrieving compiler information:
Compiler using remote version: 'v0.8.0+commit.c7dfd78e', solidity version: 0.8.0+commit.c7dfd78e.Emscripten.clang

```

error in terminal:
```
PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

C:/Users/rosne/.brownie/packages/OpenZeppelin/openzeppelin-contracts@4.3.0/contracts/access/Ownable.sol:3:1: ParserError: Source file requires different compiler version (current compiler is 0.6.12+commit.27d51765.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^

C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.2.1/contracts/src/v0.8/VRFConsumerBase.sol:2:1: ParserError: Source file requires different compiler version (current compiler is 0.6.12+commit.27d51765.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery>
```
Lottery.sol
```ruby
// SPDX-License-Identifier: MIT
pragma solidity ^0.6.0;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@chainlink/contracts/src/v0.8/VRFConsumerBase.sol";

contract Lottery is VRFConsumerBase, Ownable {
    uint256 usdEntryFee;
    address payable[] public players;
    address payable public recentWinner;
    uint256 public randomness;
    AggregatorV3Interface internal ethUsdPriceFeed;
    enum LOTTERY_STATE {
        OPEN,
        CLOSED,
        CALCULATING_WINNER
    }

    LOTTERY_STATE public lottery_state;
    uint256 public fee;
    bytes32 public keyhash;

    constructor(
        address _priceFeedAddress,
        address _vrfCoordinator,
        address _link,
        uint256 _fee,
        bytes32 _keyhash
    ) public VRFConsumerBase(_vrfCoordinator, _link) {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
        lottery_state = LOTTERY_STATE.CLOSED;
        fee = _fee;
        keyhash = _keyhash;
    }

    function enter() public payable {
        //$50 min
        require(lottery_state == LOTTERY_STATE.OPEN);
        require(msg.value >= getEntranceFee(), "Not enough ETH!");
        players.push(payable(msg.sender));
    }

    function getEntranceFee() public view returns (uint256) {
        (, int256 price, , , ) = ethUsdPriceFeed.latestRoundData();
        uint256 adjustedPrice = uint256(price) * 10**12; //18 decimals
        //$50, 2000 ETH
        //50/2000
        //50*10000/2000
        uint256 costToEnter = (usdEntryFee * 10**18) / adjustedPrice;
        return costToEnter;
    }

    function startLottery() public onlyOwner {
        require(
            lottery_state == LOTTERY_STATE.CLOSED,
            "cant start a new lottery yet"
        );
        lottery_state = LOTTERY_STATE.OPEN;
    }

    function endLottery() public onlyOwner {
        lottery_state = LOTTERY_STATE.CALCULATING_WINNER;
        bytes32 requestId = requestRandomness(keyhash, fee);
    }

    function FulfillRandomness(bytes32 _requestId, uint256 _randomness)
        internal
        override
    {
        require(
            lottery_state == LOTTERY_STATE.CALCULATING_WINNER,
            "you arent there yet!"
        );

        require(_randomness > 0, "random not found");
        uint256 indexOfWinner = _randomness % players.length;
        recentWinner = players[indexOfWinner];
        recentWinner.transfer(address(this).balance);

        //reset

        players = new address payable[](0);
        lottery_state = LOTTERY_STATE.CLOSED;
        randomness = _randomness;
    }
}
```
I also tried to google some solutions so my settings.json file is a bit different but that didnt help too
settings.json:
```ruby
{
    "solidity.compileUsingRemoteVersion": "v0.8.0+commit.c7dfd78e",
    "solidity.defaultCompiler": "remote",
    "solidity.compileUsingLocalVersion": "d:\\Python projects\\Solidity dev\\demo\\smartcontract-lottery\\soljson-v0.8.0+commit.c7dfd78e.js"
    // "solidity.compileUsingRemoteVersion": "v0.7.4+commit.3f05b770",
    // "solidity.enableLocalNodeCompiler": false
}
```
In the brownie-config.yaml, I tried all the versions of openzepplin contracts too,
brownie-config.yaml
```ruby
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
  - OpenZeppelin/openzeppelin-contracts@4.3.0
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@0.2.1'
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@4.3.0'
networks:
  mainnet-fork:
    eth_usd_price_feed: '0xaEA2808407B7319A31A383B6F8B60f04BCa23cE2'
```

Any help will be much appreciated.

**Reply by WeTheArtist** [reply]
I also tried to change the compiler with 
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

Now i get a different error.
Completely lost here :(

```
PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.8.11
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

ParserError: Source file requires different compiler version (current compiler is 0.8.11+commit.d7f03943.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
 --> C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.2.1/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol:2:1:
  |
2 | pragma solidity ^0.6.0;
  | ^^^^^^^^^^^^^^^^^^^^^^^

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> 
```

**Reply by cromewar** [reply]
@WeTheArtist, remember to restart all the terminals you have opened after changing the compiler version, also it's important to delete the `build` folder.

**Reply by WeTheArtist** [reply]
@cromewar I already did that but the error remained the same :(, I also closed the application and restarted it everytime I changed the compiler but same issue.

**Reply by cromewar** [reply]
@WeTheArtist  Ok there is a workaround you could try:

1. delete the build folder again.
2. on this options:
![image](https://user-images.githubusercontent.com/20199180/147263398-35b99090-b942-43bc-a40d-ef6af1bd9180.png)
Change both, the workspace and the global compiler version.
3. change the chainllink-brownie-contracts versions to 0.3.0 (which is actually the latest)
4. go to `C:/Users/rosne/` and delete the `.brownie` folder.
5. try to compile again.

**Reply by WeTheArtist** [reply]
I also noticed that in the brownie-config.yaml file :smartcontractkit/chainlink-brownie-contracts@1.1.1 was written differently in dependencies and remappings. However, after making both of them the same it still has the same errors. I am trying several things to try and see if I can match the versions somehow. I think I cannot proceed ahead if the code doesn't compile right?
``` ruby
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@0.2.1
  - OpenZeppelin/openzeppelin-contracts@4.3.0
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@0.2.1'
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@4.3.0'
networks:
  mainnet-fork:
    eth_usd_price_feed: '0xaEA2808407B7319A31A383B6F8B60f04BCa23cE2'
```

**Reply by cromewar** [reply]
> I also noticed that in the brownie-config.yaml file :smartcontractkit/chainlink-brownie-contracts@1.1.1 was written differently in dependencies and remappings. However, after making both of them the same it still has the same errors. I am trying several things to try and see if I can match the versions somehow. I think I cannot proceed ahead if the code doesn't compile right?
> 
> ```ruby
> dependencies:
>   - smartcontractkit/chainlink-brownie-contracts@0.2.1
>   - OpenZeppelin/openzeppelin-contracts@4.3.0
> compiler:
>   solc:
>     remappings:
>       - '@chainlink=smartcontractkit/chainlink-brownie-contracts@0.2.1'
>       - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@4.3.0'
> networks:
>   mainnet-fork:
>     eth_usd_price_feed: '0xaEA2808407B7319A31A383B6F8B60f04BCa23cE2'
> ```

Yep you won't be able to continue if the code does not compile, please refer to the instructions I posted above and see if they work for you.

**Reply by WeTheArtist** [reply]
HI @cromewar ,
Thank you, I did all the steps that you mentioned but it is still giving the same problem.

"delete the build folder again.
on this options:
image
Change both, the workspace and the global compiler version.
change the chainllink-brownie-contracts versions to 0.3.0 (which is actually the latest)
go to C:/Users/rosne/ and delete the .brownie folder.
try to compile again."

Do you think that it could be because the imports that we do are for specific versions of solidity?
I am not sure if I am thinking correctly, but is it possible that if 2 out of the 3 imports run with solidity 0.8.0 and 1 with 0.6.0?

I am very new to coding and I really like web3, so if I am asking stupid questions then please ignore it. :)

```
import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@chainlink/contracts/src/v0.8/VRFConsumerBase.sol";
````

Would uninstalling VS code and all its packages help?

**Reply by cromewar** [reply]
Not stupid questions at all, there are not stupid questions, the versions are retro compatible as far as they has this `^` symbol which indicates that the compiler should use that and any other above version available, but the references are no, so If you have changed the compiler version to ^0.8.0 on the contract, you should also change this line:

`import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";`

To this line:

`import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";`

**Reply by cromewar** [reply]
Another solution is just use the compiler on version 0.8.0 selected on vscode, but keeping the compiler version on your contract to ^0.6.0 which should work and you would have to change the import.

**Comment by WeTheArtist** [comment]
Thank you @cromewar , so when I tried the solution with changing the compiler version to ^0.8.0 and import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol"; 
the error changed:


terminal:
```
PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.8.11
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

TypeError: Function has override specified but does not override anything.
  --> contracts/Lottery.sol:71:9:
   |
71 |         override
   |         ^^^^^^^^


TypeError: Contract "Lottery" should be marked as abstract.
 --> contracts/Lottery.sol:8:1:
  |
8 | contract Lottery is VRFConsumerBase, Ownable {
  | ^ (Relevant source part starts here and spans across multiple lines).
Note: Missing implementation:
   --> C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.3.0/contracts/src/v0.8/VRFConsumerBase.sol:116:3:
    |
116 |   function fulfillRandomness(bytes32 requestId, uint256 randomness) internal virtual;
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


Warning: Visibility for constructor is ignored. If you want the contract to be non-deployable, making it "abstract" is sufficient.
  --> contracts/Lottery.sol:25:5:
   |
25 |     constructor(
   |     ^ (Relevant source part starts here and spans across multiple lines).

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> 
```
On changing the compiler on version 0.8.0 selected on vscode, but keeping the compiler version on your contract to ^0.6.0, 
So i think change as below:
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.6.0;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@chainlink/contracts/src/v0.8/VRFConsumerBase.sol";
```
and for 0.8.0 onvs code changing the Solidity: change global compiler version(remote) to 0.8.0 right? 

i get another error:

terminal:
```
PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul        
CompilerError: solc returned the following errors:

C:/Users/rosne/.brownie/packages/OpenZeppelin/openzeppelin-contracts@4.3.0/contracts/access/Ownable.sol:3:1: ParserError: Source file requires different compiler version (current compiler is 0.6.12+commit.27d51765.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^

C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.3.0/contracts/src/v0.8/VRFConsumerBase.sol:2:1: ParserError: Source file requires different compiler version (current compiler is 0.6.12+commit.27d51765.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> 
```
I really appreciate you trying to help here. 
:)

**Reply by cromewar** [reply]
This ``changing the compiler version to ^0.8.0 and import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";` is or  best way to go, now please share your Lottery contract code here, so we could check it, as the errors are actually fixable on your code.

**Reply by WeTheArtist** [reply]
Keeping my fingers crossed that there is something wrong in the code and we can fix it :)

Lottery.sol:

```ruby
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@chainlink/contracts/src/v0.8/VRFConsumerBase.sol";

contract Lottery is VRFConsumerBase, Ownable {
    uint256 usdEntryFee;
    address payable[] public players;

    address payable public recentWinner;
    uint256 public randomness;
    AggregatorV3Interface internal ethUsdPriceFeed;
    enum LOTTERY_STATE {
        OPEN,
        CLOSED,
        CALCULATING_WINNER
    }

    LOTTERY_STATE public lottery_state;
    uint256 public fee;
    bytes32 public keyhash;

    constructor(
        address _priceFeedAddress,
        address _vrfCoordinator,
        address _link,
        uint256 _fee,
        bytes32 _keyhash
    ) public VRFConsumerBase(_vrfCoordinator, _link) {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
        lottery_state = LOTTERY_STATE.CLOSED;
        fee = _fee;
        keyhash = _keyhash;
    }

    function enter() public payable {
        //$50 min
        require(lottery_state == LOTTERY_STATE.OPEN);
        require(msg.value >= getEntranceFee(), "Not enough ETH!");
        players.push(payable(msg.sender));
    }

    function getEntranceFee() public view returns (uint256) {
        (, int256 price, , , ) = ethUsdPriceFeed.latestRoundData();
        uint256 adjustedPrice = uint256(price) * 10**12; //18 decimals
        //$50, 2000 ETH
        //50/2000
        //50*10000/2000
        uint256 costToEnter = (usdEntryFee * 10**18) / adjustedPrice;
        return costToEnter;
    }

    function startLottery() public onlyOwner {
        require(
            lottery_state == LOTTERY_STATE.CLOSED,
            "cant start a new lottery yet"
        );
        lottery_state = LOTTERY_STATE.OPEN;
    }

    function endLottery() public onlyOwner {
        lottery_state = LOTTERY_STATE.CALCULATING_WINNER;
        bytes32 requestId = requestRandomness(keyhash, fee);
    }

    function FulfillRandomness(bytes32 _requestId, uint256 _randomness)
        internal
        override
    {
        require(
            lottery_state == LOTTERY_STATE.CALCULATING_WINNER,
            "you arent there yet!"
        );

        require(_randomness > 0, "random not found");
        uint256 indexOfWinner = _randomness % players.length;
        recentWinner = players[indexOfWinner];
        recentWinner.transfer(address(this).balance);

        //reset

        players = new address payable[](0);
        lottery_state = LOTTERY_STATE.CLOSED;
        randomness = _randomness;
    }
}

```

**Reply by cromewar** [reply]
Ok I was able to reproduce the problem and now I'm certain what's happening. As there a lot of changes on sol 0.8.0 we should better stay at 0.6.0, so first of all change the import and the compiler back to 0.6.0.

So where is the problem? I'm going to guess but your brownie config look like this:

```YAML
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@0.3.0
  - OpenZeppelin/openzeppelin-contracts@4.3.0
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@0.3.0'
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@4.3.0'
dotenv: .env
```

Note: you can return to 1.1.1 con chainlink-brownie-contracts.

The problem might be the openzeppelin import, as it was asking you for compiler version 0.8.0, if you have 4.3.0 is actually incorrect for this contract, the correct version for this is 3.4.0. Was I correct?

**Comment by WeTheArtist** [comment]
I reverted back all the changes so now, the code looks as below and the compiler is
```
Retrieving compiler information:
Compiler using remote version: 'v0.6.0+commit.26b70077', solidity version: 0.6.0+commit.26b70077.Emscripten.clang
```
Lottery.sol
```ruby
// SPDX-License-Identifier: MIT
pragma solidity ^0.6.0;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@chainlink/contracts/src/v0.6/VRFConsumerBase.sol";

contract Lottery is VRFConsumerBase, Ownable {
    uint256 usdEntryFee;
    address payable[] public players;

    address payable public recentWinner;
    uint256 public randomness;
    AggregatorV3Interface internal ethUsdPriceFeed;
    enum LOTTERY_STATE {
        OPEN,
        CLOSED,
        CALCULATING_WINNER
    }

    LOTTERY_STATE public lottery_state;
    uint256 public fee;
    bytes32 public keyhash;

    constructor(
        address _priceFeedAddress,
        address _vrfCoordinator,
        address _link,
        uint256 _fee,
        bytes32 _keyhash
    ) public VRFConsumerBase(_vrfCoordinator, _link) {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
        lottery_state = LOTTERY_STATE.CLOSED;
        fee = _fee;
        keyhash = _keyhash;
    }

    function enter() public payable {
        //$50 min
        require(lottery_state == LOTTERY_STATE.OPEN);
        require(msg.value >= getEntranceFee(), "Not enough ETH!");
        players.push(payable(msg.sender));
    }

    function getEntranceFee() public view returns (uint256) {
        (, int256 price, , , ) = ethUsdPriceFeed.latestRoundData();
        uint256 adjustedPrice = uint256(price) * 10**12; //18 decimals
        //$50, 2000 ETH
        //50/2000
        //50*10000/2000
        uint256 costToEnter = (usdEntryFee * 10**18) / adjustedPrice;
        return costToEnter;
    }

    function startLottery() public onlyOwner {
        require(
            lottery_state == LOTTERY_STATE.CLOSED,
            "cant start a new lottery yet"
        );
        lottery_state = LOTTERY_STATE.OPEN;
    }

    function endLottery() public onlyOwner {
        lottery_state = LOTTERY_STATE.CALCULATING_WINNER;
        bytes32 requestId = requestRandomness(keyhash, fee);
    }

    function FulfillRandomness(bytes32 _requestId, uint256 _randomness)
        internal
        override
    {
        require(
            lottery_state == LOTTERY_STATE.CALCULATING_WINNER,
            "you arent there yet!"
        );

        require(_randomness > 0, "random not found");
        uint256 indexOfWinner = _randomness % players.length;
        recentWinner = players[indexOfWinner];
        recentWinner.transfer(address(this).balance);

        //reset

        players = new address payable[](0);
        lottery_state = LOTTERY_STATE.CLOSED;
        randomness = _randomness;
    }
}

```

brownie-config.yaml:
```ruby
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
  - OpenZeppelin/openzeppelin-contracts@3.4.0
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1'
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@3.4.0'
dotenv: .env
networks:
  mainnet-fork:
    eth_usd_price_feed: '0xaEA2808407B7319A31A383B6F8B60f04BCa23cE2'
```
test_lottery.py
```ruby
# 0,01243301712
# 12000000000000000

from brownie import Lottery, accounts, config, network
from web3 import Web3

# from brownie.network import account


def test_get_entrance_fee():
    account = accounts[0]
    lottery = Lottery.deploy(
        config["networks"][network.show_active()]["eth_usd_price_feed"],
        {"from": account},
    )

    # assert lottery.getEntraceFee() > Web3.toWei(0.011, "ether")
    # assert lottery.getEntraceFee() < Web3.toWei(0.015, "ether")
```

So I changed the brownie-config.yaml file with 1.1.1 for chainlink-brownie-contracts and openzeppelin import to 3.4.0 (as in the video).

So at the moment, I am getting a different error in the terminal after running brownie compile
Terminal:
```
PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

contracts/Lottery.sol:71:9: TypeError: Function has override specified but does not override anything.
        override
        ^------^

contracts/Lottery.sol:8:1: TypeError: Contract "Lottery" should be marked as abstract.
contract Lottery is VRFConsumerBase, Ownable {
^ (Relevant source part starts here and spans across multiple lines).
C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@1.1.1/contracts/src/v0.6/VRFConsumerBase.sol:121:3: Missing implementation:
  function fulfillRandomness(bytes32 requestId, uint256 randomness)
  ^ (Relevant source part starts here and spans across multiple lines).

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery>
```

**Reply by cromewar** [reply]
Yes the fullfilRandomness is an abstract function which should be overrided in order to work properly, you will do that after during the course so for now just remove the  `override` keyword from that function.

**Comment by WeTheArtist** [comment]
Yes i tried that as well, the error changed , now when i compile:

```

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

contracts/Lottery.sol:8:1: TypeError: Contract "Lottery" should be marked as abstract.
contract Lottery is VRFConsumerBase, Ownable {
^ (Relevant source part starts here and spans across multiple lines).
C:/Users/rosne/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@1.1.1/contracts/src/v0.6/VRFConsumerBase.sol:121:3: Missing implementation:
  function fulfillRandomness(bytes32 requestId, uint256 randomness)
  ^ (Relevant source part starts here and spans across multiple lines).

PS D:\Python projects\Solidity dev\demo\smartcontract-lottery> 
```

**Reply by cromewar** [reply]
My bad as they are abstract you must 100% override the function otherwise it would complain saying it should be abstract too, so please delete the build folder and copy and paste this:

```Solidity
function fulfillRandomness(bytes32 _requestId, uint256 _randomness)
        internal
        override
    {
        require(
            lottery_state == LOTTERY_STATE.CALCULATING_WINNER,
            "You aren't there yet!"
        );
        require(_randomness > 0, "random-not-found");
        uint256 indexOfWinner = _randomness % players.length;
        recentWinner = players[indexOfWinner];
        recentWinner.transfer(address(this).balance);
        // Reset
        players = new address payable[](0);
        lottery_state = LOTTERY_STATE.CLOSED;
        randomness = _randomness;
    }
```

**Reply by WeTheArtist** [reply]
Thanks A TON @cromewar , this worked! 
Was it because of the FulfillRandomness instead of fulfillRandomness at the end? So all the trouble was because of one capital F ?

**Reply by cromewar** [reply]
Yes it was, really hard to find tough. But I'm glad is solved now.

**Reply by neerajnerlekar** [reply]
Thanks for this @cromewar @WeTheArtist , even I had misspelled 'fulfillRandomness' which was giving similar errors.

**Comment by MANaSexp** [comment]
> I started to get compile error after adding openzepplin ownable contract. I'm really confused and can't find solution.
> 
> Could you please someone help me about it?
> 
> ![Screen Shot 2021-12-08 at 17 19 01](https://user-images.githubusercontent.com/18116965/145224791-093e6b49-5daa-4c55-a5ab-d4830e6412fd.png)
> 
> ![Screen Shot 2021-12-08 at 17 31 25](https://user-images.githubusercontent.com/18116965/145226022-b2c821ea-ccf2-4ab6-9d88-4e0cd19a1b3b.png)
> 
> ![Screen Shot 2021-12-08 at 17 19 33](https://user-images.githubusercontent.com/18116965/145224786-f59512ee-a2e5-4bb7-b75f-edf7e7ddc934.png)



Add following brownie-config.yaml file to your project:

compiler:
  solc:
    version: 0.6.12

When you compile, brownie will download solc version from config.
